samples/stream_download_percentage/src/sample/stream_download_percentage.cljd

```diff
@@ -0,0 +1,164 @@
+            ["dart:core" :as dc]
+            ["dart:io" :as dio]
+            ["package:http/http.dart" :as http]
+            ["package:validators/validators.dart" :as validators]
+            [cljd.flutter :as f]
+            [clojure.string :as str]))
+
+(defn calculate-percentage
+  [{::keys [total
+            completion]}]
+  (if (= 0 total)
+    0
+    (-> (- total
+           completion)
+        (/ total)
+        (* 100)
+        (- 100)
+        (* -1))))
+
+
+(defn delete-local-file!
+  [file-name]
+  (let [old-file (-> file-name
+                     dio/File)]
+    (when (await (.exists old-file))
+      (-> old-file
+          .delete))))
+
+
+(defn parse-uri
+  [tc]
+  (let [form-text (-> tc
+                      .-text)]
+    (when (validators/isURL form-text)
+      (-> form-text
+          Uri/parse))))
+
+
+(defn start-button-handler
+  [{::keys [file-name
+            request-uri
+            *total
+            *completion
+            file-image-atom]}]
+  (delete-local-file! file-name)
+  (let [streamed-response (some->> request-uri
+                                   (http/Request "get")
+                                   .send
+                                   await)
+        local-file-open (dio/File file-name)
+        opened-file (.openWrite local-file-open)
+        _total-bytes-add (swap! *total
+                                #(+ % (.-contentLength streamed-response)))]
+    (some-> streamed-response
+            .-stream
+            (.listen (fn [event]
+                       (swap! *completion #(+ % (.-length event)))
+                       (.add opened-file event))
+                     .onDone (fn []
+                               (reset! file-image-atom local-file-open))))))
+
+(defn show-image-ui
+  [{::keys [file-image]}]
+  (f/widget
+    :watch [file-exists? (some-> file-image
+                                 .exists)]
+    m/Center
+    (if file-exists?
+      (-> file-image
+          .readAsBytesSync
+          (m/Image.memory
+            .errorBuilder (fn [_ _ _]
+                            (m/Text "Try to download a valid image file"))))
+      (m/Text "Add a link to download an image to show here"))))
+
+
+(defn show-image-download-button-ui
+  [{::keys [file-name
+            path-str]}]
+  (f/widget
+    :managed [tc (m/TextEditingController)]
+    :let [{:flds [text]} tc]
+    :watch [file-image (atom (dio/File. file-name)) :as file-image-atom]
+    m/Column
+    .children
+    [(m/Spacer)
+     (show-image-ui {::file-image file-image})
+     (m/Spacer)
+     (f/widget
+       (m/TextField
+         .decoration (m/InputDecoration
+                       .border (m/OutlineInputBorder)
+                       .hintText "Add the link to file download")
+         .controller tc))
+     (f/widget
+       (m/Row .mainAxisAlignment m/MainAxisAlignment.center)
+       .children
+       [(m/Spacer)
+        (f/widget
+          :let [*completion (atom 0)
```

Comment:
Never create atoms in `:let` or anything stateful, they could be recreated by accident. use `:watch :as` (or `:managed :dispose nil` if you don't watch on them immediately).
Matter of style but I'd rather have a single atom holding completion a total than two.